internal/testdata/plugins/java.lua

```diff
@@ -0,0 +1,180 @@
+            }
+        }
+    }
+end
+
+--- Extension point, called after PreInstall, can perform additional operations,
+--- such as file operations for the SDK installation directory or compile source code
+--- Currently can be left unimplemented!
+function PLUGIN:PostInstall(ctx)
+    --- ctx.rootPath SDK installation directory
+    local rootPath = ctx.rootPath
+    local runtimeVersion = ctx.runtimeVersion
+    local sdkInfo = ctx.sdkInfo['sdk-name']
+    local path = sdkInfo.path
+    local version = sdkInfo.version
+    local name = sdkInfo.name
+end
+
+--- Return all available versions provided by this plugin
+--- @param ctx table Empty table used as context, for future extension
+--- @return table Descriptions of available versions and accompanying tool descriptions
+function PLUGIN:Available(ctx)
+    local runtimeVersion = ctx.runtimeVersion
+    return {
+        {
+            version = "xxxx",
+            note = "LTS",
+            addition = {
+                {
+                    name = "npm",
+                    version = "8.8.8",
+                }
+            }
+        }
+    }
+end
+
+--- Each SDK may have different environment variable configurations.
+--- This allows plugins to define custom environment variables (including PATH settings)
+--- Note: Be sure to distinguish between environment variable settings for different platforms!
+--- @param ctx table Context information
+--- @field ctx.path string SDK installation directory
+function PLUGIN:EnvKeys(ctx)
+    --- this variable is same as ctx.sdkInfo['plugin-name'].path
+    local mainPath = ctx.path
+    local runtimeVersion = ctx.runtimeVersion
+    local sdkInfo = ctx.sdkInfo['sdk-name']
+    local path = sdkInfo.path
+    local version = sdkInfo.version
+    local name = sdkInfo.name
+    return {
+        {
+            key = "JAVA_HOME",
+            value = mainPath
+        },
+        {
+            key = "PATH",
+            value = mainPath .. "/bin"
+        }
+    }
+end
+
+--- When user invoke `use` command, this function will be called to get the
+--- valid version information.
+--- @param ctx table Context information
+function PLUGIN:PreUse(ctx)
+    local runtimeVersion = ctx.runtimeVersion
+    --- user input version
+    local version = ctx.version
+    --- installed sdks
+    local sdkInfo = ctx.installedSdks['xxxx']
+    local path = sdkInfo.path
+    local name = sdkInfo.name
+    local sdkVersion = sdkInfo.version
+
+    --- working directory
+    local cwd = ctx.cwd
+
+    PrintTable(ctx)
+
+    --- user input scope
+    local scope = ctx.scope
+
+    if (scope == "global") then
+        print("return 9.9.9")
+        return {
+            version = "9.9.9",
+        }
+    end
+
+    if (scope == "project") then
+        print("return 10.0.0")
+        return {
+            version = "10.0.0",
+        }
+    end
+
+    print("return 1.0.0")
+
+    return {
+        version = "1.0.0"
+    }
+end
+
+function PrintTable(t, indent)
```

Comment:
could we add this function into builtin function?

printTable is very useful when debug

Reply:
Of course!

Reply:
Of course!